Add developer documentation and improve README

jrwdunham opened the pull request:
Fixes #83

sevein commented:
Nicely done @jrwdunham!

jrwdunham commented:
@ross-spencer @joel-simpson I'm tagging you two here because I know you have been writing a lot of Gherkin recently so I'm anticipating you may have comments about this PR or may even want to make suggestions (or add new commits) about guidelines that we should be providing for how to write feature files.

ross-spencer commented on README.rst:
```diff
@@ -1,90 +1,142 @@
-Consider the following snippet from the premis-events.feature file::
+This repository contains automated user acceptance tests for Archivematica_
+(AM) written using the Python behave_ library and the Gherkin_ language. Using
+Gherkin to express tests makes them readable to a wide range of Archivematica
```
Maybe we can add a line that describes this as a technique from Behavioural Driven Development which focus on what a user wants a system to do, not how it does it.

joel-simpson replied:
I think the Behave documentation is great for describing how all these things fit together and even has a section on BDD. So maybe we could add a link to that and say something like:
"The Behave documentation provides a good overview of the key concepts and their origins in behaviour driven development." http://behave.readthedocs.io/en/latest/

jrwdunham replied:
I've added both of these things in a footnote. Does it work?

ross-spencer replied:
:+1:

ross-spencer commented on README.rst:
```diff
@@ -1,90 +1,142 @@
+
+The AMAUAT are a completely separate application from `Archivematica`_ (AM) and
+the Archivematica `Storage Service`_ (SS). They require that you already have an
+Archivematica instance deployed somewhere to test against (see `Installing
```
Suggested:

> They require that you already have an Archivematica instance deployed somewhere that you can test against

ross-spencer commented on README.rst:
```diff
@@ -1,90 +1,142 @@
+AIPs being created in the AM instance that is being tested.
+
+The tests use Selenium_ to launch a browser in order to interact with
+Archivematica's web interfaces. Therefore, you may need to install a web
```
Suggested: 

> The tests use [Selenium WebDriver]() to launch a web-browser in order to interact with Archivematica's web interfaces.

ross-spencer commented on README.rst:
```diff
@@ -1,90 +1,142 @@
+Installation quickstart
+--------------------------------------------------------------------------------
+
+The following list of commands illustrates the bare minimum required in order
```
Redundant pluralisation: `commands illustrate`

jrwdunham replied:
No, cuz it's singular `list` that is the subject of the sentence, right?

ross-spencer commented on README.rst:
```diff
@@ -1,90 +1,142 @@
-A browser (Chrome or Firefox) must be installed on the system where the tests
-are being run; see below. On a dev or CI server, this may require installation.
+Archivematica's public Vagrant/Ansible deployment tool `deploy-pub`_ allows you
+to install the AMAUAT when provisioning your VM. This simply requires setting
```
First instance of Virtual Machine, define the acronym here (VM).

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+assertions as specified in human-readable Gherkin_ "feature" files.
+
+Gaining a thorough understanding of the tests requires understanding these three
+layers:
```
I feel like "To understand Automated Acceptance Testing" is too general. We're not explaining automated acceptance testing in general, we're explaining how this particular application does automated acceptance testing.

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+   files (see ``features/steps/``).
+3. `The Archivematica User package`_: the ``ArchivematicaUser`` class which has
+   the ability to interact with Archivematica via its browser-based GUIs and
+   its APIs.
```
good idea

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+--------------------------------------------------------------------------------
+
+The feature files describe features of Archivematica. Each feature file defines
+a ``Feature`` by declaring one or more scenarios. These ``Scenario`` objects
```
It is an object in the programming sense, in that in the steps files you interact with it like `context.scenario.sip_uuid = ...`, etc. However, I think that's confusing things when we're describing feature files, so I'll remove the "object". I've also removed the "These" but kept "represent".

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+
+    Given+ (When+ Then+)+
+
+For more information on how to write Gherkin, see this Gherkin_ page and this
```
My thought is that talking about Cucumber would just be distracting and is unnecessary. Isn't Cucumber (given that it is an implementation) more comparable to Behave than it is to Gherkin? It would be like documenting a HTTP library in Python and then talking about a comparable library in GoLang, would it not?

ross-spencer replied:
@jrwdunham a mistake on my part. Sorry about that.

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+     an appropriately named subdirectory of ``features/``, e.g.,
+     ``features/CCA/``.
+
+3. Tags. Feature files contain tags so that users running the ``behave``
```
Good point about defining tags, indicating their optionality, and distinguishing Python decorators. I've also sub-itemized the two special tags and their descriptions.

I don't think we need to formalize the addition of new tags. If a PR comes in that uses a new tag, then we evaluate that during CR. I also don't think we want to formalize a closed class of special tags: that seems like regulatory overkill at this point to me.

joel-simpson replied:
I'm not sure we need to mention the Artefactual stylistic decision (not that it wouldn't be interesting to some readers, but I think less is more to keep the document easy to read). 
I like your suggestion of lists for the tags. Perhaps we could have two sections: 
Feature specific tags:

- Each feature should have.... 
Common tags: 

- @wip tag is used for....

- @non-executable is used for... 

And don't we have some other common tags? e.g. for version?

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+     including flags, as well as details of the type of Archivematica deploy(s)
+     that the behave command was successfully run against.
+
+6. Existing steps. Whenever possible, new feature files should use existing
```
I think emboldening is good, but I'm standing firm on the punctuation here!

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+Steps files
+================================================================================
+
+What are steps files?
```
Why should they be capitalized here? I'm consistently doing sentence-style capitalization on all headers here, so why capitalize "steps" here?

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+    @given('the default processing config is in its default state')
+    def step_impl(context):
+        ...
+
```
`Given the default processing config is in its default state` is a step. It is a "`Given` step" or a "`Given`-type step". Isn't this sufficiently described in the "What are feature files?" subsection?

What you say isn't quite correct. The `step_impl` function doesn't really "run the scenario". Each step in a feature file has a direct mapping to a single function named `step_impl`. That mapping is made by the text of the step exactly matching the string argument to the appropriate Python decorator.

I'm not saying that the documentation here couldn't be made more clear. I'm just not sure quite how to do that. More suggestions?

joel-simpson replied:
rightly or wrongly (I don't know the answer), the Behave documentation treats Given (as well as When and Then) as steps, so I don't see a problem with the way this is worded. http://behave.readthedocs.io/en/latest/tutorial.html#feature-files

Also speaking as a non-dev (I forgot to mention this above), I don't know anything about decorators in python. Distinguishing tags from decorators wouldn't help me, nor would your comment about thinking about given as a decorator.

jrwdunham replied:
I've added "Note that behave tags are completely distinct from Python's decorator syntax, which is superficially similar in that it too uses the @ character as a prefix." I'm hoping that is more helpful than distracting. Maybe it should be in parens? Or in a footnote?

joel-simpson commented on README.rst:
```diff
@@ -1,90 +1,142 @@
-Using Gherkin to express tests makes them quite readable to non-programmers.
-Consider the following snippet from the premis-events.feature file::
+This repository contains automated user acceptance tests for Archivematica_
+(AM) written using the Python behave_ library and the Gherkin_ language. Using
```
Don't think we need to mention cucumber - it is an alternative to Behave. See tool comparison here: http://behave.readthedocs.io/en/latest/comparison.html

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+How to create new steps
+--------------------------------------------------------------------------------
+
+If you need to create a step in a feature file that is not yet implemented as a
```
I've added more details to the above section to hopefully clarify this.

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+
+Functions that do not implement steps but which are used by step functions
+should be defined in ``features/steps/utils.py`` and imported into the step
+modules as needed.
```
I think I'll make it a link to the module, but they're so varied that I'm not sure an example would be very illuminating.

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+
+The step functions described in the section above can access the
+``ArchivematicaUser`` instance using the ``am_user`` attribute of the
+``context`` object. For example, in the step function for ``When the user
```
I don't think so. It's a code literal, so I'm writing it as it's actually used, i.e., as `context.am_user.api.download_aip(...)`.

ross-spencer replied:
Understood.

We don't describe specifically what 'context' is anywhere. It makes sense up to a point as it's not an object we create - but it is specific to behave and we do reference it. Should we describe that object (http://behave.readthedocs.io/en/latest/api.html#detecting-that-user-code-overwrites-behave-context-attributes) (a single-line) or put that link somewhere just to explain to folk a bit more why it's the first argument passed to every step function? Maybe https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/84/files#diff-8840386bbef65a3752f8040bc35a81beR218

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+Configuration
+================================================================================
+
+The file at ``features/environment.py`` defines a ``before_scenario`` function
```
k

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+================================================================================
+
+The file at ``features/environment.py`` defines a ``before_scenario`` function
+which is a hook that behave_ calls before each scenario is run. Each time this
```
I did it, but I'm still unsure. The [behave](https://behave.readthedocs.io/en/latest/) docs seems to use `behave` quite a bit. It's confusing when you have a CL tool in lowercase...

jrwdunham commented on docs/developer-documentation.rst:
```diff
@@ -0,0 +1,486 @@
+   running on the previously created SIP/AIP.
+
+
+How do I run the tests of the tests?
```
Good ideas. I've added drafts of these sections.

> So should we (do we?) always have a reference server available and configured so that any of us, analysts, devops, etc. can just press Go and let them run?

Maybe. One can always just run the tests against any AM instance. Of course, that requires a bit of technical know-how. It's not a single button press kind of thing yet.

It would be nice to promise something about how we run the tests before a release and how exactly we do that. The problem is that at present this is still very informal. Also, it is complicated by the fact that we are not versioning these tests yet. I think we need to start versioning and then be able to say: Version A of Archivematica should be expected to pass all of the tests T of version B of AMAUAT. In short, I think we need to do more work on the tests and on our CI before we can promise anything like that here.

ross-spencer replied:
Awesome! They look great! Thanks for your additional insights as well. Something we can aim for! :)

jrwdunham commented:
@ross-spencer thanks for all the detailed feedback. I think I've implemented all your suggested changes or responded to them. Pong to you to approve or continue the refinement.

@joel-simpson thanks for your comments too, which I think I've incorporated now too. I agree about your idea of adding a third style guide document. A first draft would be great. I'm thinking that would be a separate PR addressing a new Issue? Do you agree? If so, creating that issue would be the first step.

joel-simpson commented:
Great suggestion @jrwdunham - I will create a new issue first and describe some of the challenges I think we should try to address with a style guide.

jrwdunham commented:
@ross-spencer I've made another set of small changes. Can you formally approve (or respond)? Thanks!
